text/0000-unsafe-aliased.md

```diff
@@ -0,0 +1,399 @@
+`impl UnsafeUnpin for T` would add the `noalias` back to `&mut T`, and that can be unsound.
+(See [the `poll_fn` debacle](https://internals.rust-lang.org/t/surprising-soundness-trouble-around-pollfn/17484).)
+With a time machine we'd make `Unpin` unsafe;
+in lieu of that, we can add this unsafe trait unstably so that we can at least resolve the existing soundness trouble while we try to figure out if it worth transitioning `Unpin` to an unsafe trait somehow.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The proposal in this RFC matches [what was discussed with the lang team a long time ago](https://github.com/rust-lang/rust/issues/63818#issuecomment-526579977).
+
+However, of course one could imagine alternatives:
+
+- Keep the status quo. The current sitaution is that we only make aliasing requirements on mutable references if the type they point to is `Unpin`. This is unsatisfying: `Unpin` was never meant to have this job. A consequence is that a stray `impl Unpin` on a `Wrapper<T>`-style type can [lead to subtle miscompilations](https://internals.rust-lang.org/t/surprising-soundness-trouble-around-pollfn/17484/15) since it re-adds aliasing requirements for the inner `T`.
+  Contrast this with the `UnsafeCell` situation, where it is not possible for (stable) code to just `impl Freeze for T` in the wrong way -- `UnsafeCell` is *always* recognized by the compiler.
+
+  On the other hand, `UnsafePinned` is rather quirky in its behavior and having two marker traits (`UnsafeUnpin` and `Unpin`) might be too confusing, so sticking with `Unpin` might not be too bad in comparison.
+
+  If we do that, however, it seems preferrable to transition `Unpin` to an unsafe trait. There *is* a clear statement about the types' invariants associated with `Unpin`, so an `impl Unpin` already comes with a proof obligation. It just happens to be the case that in a module without unsafe, one can always arrange all the pieces such that the proof obligation is satisifed.
+  This is mostly a coincidence  and related to the fact that we don't have safe field projections on `Pin`. That said, solving this also requires solving the trouble around `Drop` and `Pin`, where effectively an `impl Drop` does an implicit `get_mut_unchecked`, i.e., implicitly assumes the type is `Unpin`.
+- `UnsafePinned` could affect aliasing guarantees *both* on mutable and shared references. This would avoid the currently rather subtle situation that arises when one of many aliasing `&mut UnsafePinned<T>` is cast or coerced to `&UnsafePinned<T>`: that is a read-only shared reference and all aliases must stop writing!
+  It would make this type strictly more 'powerful' than `UnsafeCell` in the sense that replacing `UnsafeCell` by `UnsafePinned` would always be correct. (Under the RFC as written, `UnsafeCell` and `UnsafePinned` can be nested to remove aliasing requirements from both shared and mutable references.)
```

Comment:
What is the argument for `UnsafePinned` affecting only aliasing on `&mut`? More optimizations? Why would I want to have `&UnsafePinned` and not have the same behavior as `&UnsafeCell`? Not sure if I missed this above, but if there is no real usecase/motivation, why have this footgun?

The problem is also that this persists if we consider `Pin<&mut UnsafePinned<T>>`, since [`into_ref`](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.into_ref) exists and `Pin<P>` always implements `Deref` if `P: Deref`.

I do not want to end up in a situation where I have to ensure I never coerce a `&mut UnsafePinned` -- or worse, a `&mut T` with a nested `UnsafePinned` -- to `&UnsafePinned`. This makes it essentially unsound to implement traits with methods that take `&self` for types that contain `UnsafePinned`.

In RFL I think all our use cases of `UnsafePinned` would then actually use `UnsafePinned<UnsafeCell>` (or swapped, whatever would be the correct/expected way).

Reply:
> In RFL I think all our use cases of UnsafePinned would then actually use UnsafePinned<UnsafeCell> (or swapped, whatever would be the correct/expected way).

Yeah @Darksonn mentioned that as well. For interior linked lists this is needed since someone might have an `&` reference to a node in the linked list and now if someone else removes an adjacent list element, these pointers will get updated.

However, for futures it is not needed since when there is an `&` reference to the future we know the future cannot currently be polled so the aliases that point to the pinned element are also all inert right now. At least I think that's true? Or could it be that a pointer to the pinned field has escaped somewhere else where it is being used even if the future is inert? I don't think that can ever happen, after all if the future is currently paused I might just drop it. I do have to call its destructor for that (that's the `drop` guarantee) but I don't think that will do anything.

I tend to agree with you though, we specifically added the `drop` guarantee to enable the use of pinning for intrusive collections, it only seems fair to also support them by accounting for the mutability that arises from them. An `&` reference to a future is pretty much useless so it's not like we're losing out on optimizations by losing the `noalias` there.

Reply:
I guess potentially `Stream::size_hint` taking `&self` could be an example where it might make a difference?

Reply:
Yeah true once those types have other methods, the `noalias` on a shared reference could be useful.